src/flamenco/gossip/fd_gossip.c

```diff
@@ -1477,45 +1506,57 @@ fd_gossip_refresh_push_states( fd_gossip_t * glob, fd_pending_event_arg_t * arg
   }
 
   /* Delete states which no longer have active peers */
-  for (ulong i = 0; i < glob->push_states_cnt; ++i) {
-    fd_push_state_t* s = glob->push_states[i];
+  ulong idx = 0;
+  ulong delete[FD_PUSH_LIST_MAX];
+  ulong delete_cnt = 0;
+  for ( fd_push_state_deque_iter_t iter = fd_push_state_deque_iter_init( glob->push_state_deque );
+        !fd_push_state_deque_iter_done( glob->push_state_deque, iter );
+        iter = fd_push_state_deque_iter_next( glob->push_state_deque, iter ), idx++ ) {
+    fd_push_state_t const * s = fd_push_state_deque_iter_ele_const( glob->push_state_deque, iter );
     if (fd_active_table_query(glob->actives, &s->addr, NULL) == NULL) {
-      fd_valloc_free(glob->valloc, s);
-      /* Replace with the one at the end */
-      glob->push_states[i--] = glob->push_states[--(glob->push_states_cnt)];
+      delete[delete_cnt++] = idx;
     }
   }
-  if (glob->push_states_cnt == FD_PUSH_LIST_MAX) {
+  for ( ulong i = 0UL; i < delete_cnt; i++ ) {
+    fd_push_state_deque_pop_idx_tail( glob->push_state_deque, delete[i] );
```

Comment:
I really don't like using a deque here. The push states are fairly large, and pop_idx_tail copies them around. deque isn't the natural data structure anyway. Use an array of pointers like before, but draw the elements from a fd_pool instead of allocing them.

Reply:
hmm i thought it was just updating pointers internally, changing to fd_pool